modules/gapi/src/backends/ocl/goclcore.cpp

```diff
@@ -523,6 +541,47 @@ GAPI_OCL_KERNEL(GOCLTranspose, cv::gapi::core::GTranspose)
     }
 };
 
+GAPI_OCL_KERNEL(GOCLBGR, cv::gapi::streaming::GBGR)
+{
+    static void run(const cv::MediaFrame& in, cv::UMat& out)
+    {
+        (void)in; (void)out;
+#ifdef HAVE_DIRECTX
+#ifdef HAVE_D3D11
+#ifdef HAVE_ONEVPL
+        auto d = in.desc();
+        GAPI_Assert(d.fmt == cv::MediaFormat::NV12);
+
+        // FIXME: consider a better solution.
+        // Current approach cannot be easily extended for other adapters (getHandle).
+        auto params = in.get<cv::gapi::wip::onevpl::VPLMediaFrameDX11Adapter>()->getHandle();
+        auto handle = cv::util::any_cast<mfxHDLPair>(params);
+        ID3D11Texture2D* texture = reinterpret_cast<ID3D11Texture2D*>(handle.first);
+        GAPI_Assert(texture != nullptr && "mfxHDLPair contains ID3D11Texture2D that is nullptr");
+
+        ID3D11Device* pD3D11Device = nullptr;
+        texture->GetDevice(&pD3D11Device);
+        GAPI_Assert(pD3D11Device != nullptr && "D3D11Texture2D::GetDevice returns pD3D11Device that is nullptr");
+
+        // FIXME: Initializing context in every run might degrade the performance
+        cv::ocl::Context& ctx = cv::directx::ocl::initializeContextFromD3D11Device(pD3D11Device);
+        GAPI_Assert(ctx.ptr() != nullptr && "Built without D3D11 support");
+
+        // FIXME: copying memory here!
+        // Not sure if it's possible without a copy here since
+        // there is no known way to convert ID3D11Texture2D to OCL memory (UMat),
+        // despite those 2 are both located (out is not allocated yet?) on the same GPU
+        cv::directx::convertFromD3D11Texture2D(texture, out);
```

Comment:
This is the second major issue. There is a copy here (+1 more copy in OCL backend at the end of execution).

Reply:
Why is it copying? looks like it must translate descriptors on already allocated data

Reply:
But in case `NV12` copy is inevitably anyway, isn't it?

Reply:
I see the name of function is 'convertFromD3Dtexture2D' which means no convert to BGR from NV12, doesn't it? But if such conversion has taken place ( especially with copy) then it means, at least, that function name is incorrect and confusing

Reply:
Can this discussion be closed now?

Reply:
what's about function 'convertFromD3Dtexture2D' renaming to reflect its purpose?
if it is not possible - then clearance comment is required at least: because there are no simple word about NV12 or other stuff which happen here

Reply:
It comes from `cv::directx` so we cannot change that. I'll add a comment what it does inside